util/types/time.go

```diff
@@ -471,6 +471,14 @@ func (t *Time) Sub(t1 *Time) Duration {
 	}
 }
 
+// Add adds d to t, returns a duration value.
+// Note that add should not be done on different time types.
+func (t *Time) Add(d *Duration) Duration {
+	d.Duration = gotime.Duration(-int64(d.Duration))
+	t2, _ := d.ConvertToTime(t.Type)
```

Comment:
There are so many things that should be taken into consideration:

* datetime contains 0
* check date out range
* different fsp 
* error handling 

It's hardly do it 'right', there are some way:

* use go standard library
* use the same implementation as mysql

Maybe this function would help https://github.com/pingcap/tidb/blob/master/util/types/mytime.go#L282